### PR TITLE
PoC of verifier for UAST

### PR DESCRIPTION
**Pull request by saberduck**
*DO_NOT_MERGE*

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/uast/UastSpec.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    properties.add(hasChild);
+  }
+
+  List<Property> build() {
+    return properties;
+  }
+}
+
+
+class UastVerifier {
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/uast/UastSpec.java:68]: Code Smell Code Smell: Add a private constructor to hide the implicit public one. ([squid:S1118](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1118))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=248&issues=AWLT92XQIx9En2YzdAca&open=AWLT92XQIx9En2YzdAca)

[ISSUE_KEY]: AWLT92XQIx9En2YzdAca

**Review by SonarTech — COMMENTED** [review]
SonarQube analysis found issues:
Bug Bugs: 0
Vulnerability Vulnerabilities: 0
Code Smell Code Smells: 4

Including the following issue(s) which could not be reported in line:
1. Code Smell Code Smell: There are 5 top-level types in this file; move all but one of them to other files. ([more](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=248&issues=AWLT92XQIx9En2YzdAcY&open=AWLT92XQIx9En2YzdAcY))

[See all issues in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=248&resolved=false)

File: common-rule-engine/src/main/java/org/sonar/uast/UastSpec.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+  static UastSpec spec() {
+    return new UastSpec();
+  }
+
+  List<Property> properties() {
+    return b.build();
+  }
+
+  void SWITCH() {
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/uast/UastSpec.java:42]: Code Smell Code Smell: Rename this method name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. ([squid:S00100](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00100))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=248&issues=AWLT92XQIx9En2YzdAcb&open=AWLT92XQIx9En2YzdAcb)

[ISSUE_KEY]: AWLT92XQIx9En2YzdAcb

File: common-rule-engine/src/main/java/org/sonar/uast/UastSpec.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      throw new IllegalStateException("Property " + this + " not fulfilled " + node);
+    }
+    if (node.is(parent) && node.getChild(child).isPresent()) {
+      tested = true;
+    }
+  }
+
+  @Override
+  public boolean tested() {
+    // TODO when optional is true we should test for both positive and negative case
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/uast/UastSpec.java:122]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=248&issues=AWLT92XQIx9En2YzdAcZ&open=AWLT92XQIx9En2YzdAcZ)

[ISSUE_KEY]: AWLT92XQIx9En2YzdAcZ

**Comment by saberduck** [comment]
This idea evolved into validators, so closing the PR

**Pull request closed without merging**